### PR TITLE
IGNITE-14708: Add an ability to track request handling completion in …

### PR DESCRIPTION
**Pull request by Sega76**
…GridRestProcessor

Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [x] There is a single JIRA ticket related to the pull request. 
- [x] The web-link to the pull request is attached to the JIRA ticket.
- [x] The JIRA ticket has the _Patch Available_ state.
- [x] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [x] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [x] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/rest/GridRestProcessor.java [hunk 1]
```diff
@@ -145,14 +146,22 @@
...
         @Override public GridRestResponse handle(GridRestRequest req) throws IgniteCheckedException {
             return handleAsync(req).get();
         }
 
         @Override public IgniteInternalFuture<GridRestResponse> handleAsync(GridRestRequest req) {
-            return handleAsync0(req);
+            IgniteInternalFuture<GridRestResponse> fut = handleAsync0(req);
+
+            if (lsnr != null)
+                fut.listen((f) -> lsnr.accept(req, f));
```
Comment (andrey-kuznetsov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/rest/GridRestProcessor.java:162]: Extra parentheses around f

File: modules/rest-http/src/test/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/RestProcessorListenersTest.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+        GridRestProcessor rest = (GridRestProcessor)ignite.context().rest();
+        rest.listener(map::put);
+
+        executeCommand(GridRestCommand.VERSION);
+
+        assertFalse(map.isEmpty());
+
+        Map.Entry<GridRestRequest, IgniteInternalFuture<GridRestResponse>> entry =
+            new ArrayList<>(map.entrySet()).get(0);
```
Comment (andrey-kuznetsov) [inline modules/rest-http/src/test/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/RestProcessorListenersTest.java:72]: No need for ArrayList creation, just use iterator().

File: modules/rest-http/src/test/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/RestProcessorListenersTest.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import org.apache.ignite.internal.processors.rest.request.GridRestRequest;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.junit.Test;
+
+/**
+ * Tests REST processor listener.
+ */
+public class RestProcessorListenersTest extends GridCommonAbstractTest {
+    /** */
+    private final Map<GridRestRequest, IgniteInternalFuture<GridRestResponse>> map = new HashMap<>();
```
Comment (andrey-kuznetsov) [inline modules/rest-http/src/test/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/RestProcessorListenersTest.java:45]: Can be turned into local variable; also, concurrent map would look safer.

**Review by andrey-kuznetsov — CHANGES_REQUESTED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/rest/GridRestProcessor.java (hunk 1 above)
Comment (Sega76) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/rest/GridRestProcessor.java]: fixed

File: modules/rest-http/src/test/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/RestProcessorListenersTest.java (hunk 2 above)
Comment (Sega76) [reply modules/rest-http/src/test/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/RestProcessorListenersTest.java]: done

File: modules/rest-http/src/test/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/RestProcessorListenersTest.java (hunk 3 above)
Comment (Sega76) [reply modules/rest-http/src/test/java/org/apache/ignite/internal/processors/rest/protocols/http/jetty/RestProcessorListenersTest.java]: done

**Review by andrey-kuznetsov — APPROVED** [review]

File: modules/core/src/main/java/org/apache/ignite/internal/processors/rest/GridRestProcessor.java [hunk 4]
```diff
@@ -145,6 +148,10 @@
     /** */
     private final Thread sesTimeoutCheckerThread;
 
+    /** Request / response listeners. */
+    private final List<BiConsumer<GridRestRequest, IgniteInternalFuture<GridRestResponse>>> lsnrs =
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/rest/GridRestProcessor.java:152]: All we need is made `handleAsync0` protected and extend `GridRestProcessor` in the plugin like this:

```
    /** */
    private static class MyGridRestProcessor extends GridRestProcessor {
        /** @param ctx Context. */
        public MyGridRestProcessor(GridKernalContext ctx) {
            super(ctx);
        }

        /** {@inheritDoc} */
        @Override protected IgniteInternalFuture<GridRestResponse> handleAsync0(GridRestRequest req) {
            IgniteInternalFuture<GridRestResponse> fut = super.handleAsync0(req);

            // Do whatever you want.

            return fut;
        }
    }
```

**Review by nizhikov — APPROVED** [review]

**Pull request merged**